src/full/Agda/Compiler/Malfunction.hs

```diff
@@ -0,0 +1,247 @@
+        Primitive{ primName = s } -> return $ Just $ Left $ Left (q, s)
+        Axiom{}                   -> return $ Just $ Left $ Right q
+        _                         -> return Nothing
+
+compile :: Mlf.Env -> [(QName, TTerm)] -> Mod
+compile env bs = Mlf.compile env bs
+
+qnamesInTerm :: Set QName -> TTerm -> Set QName
+qnamesInTerm s t = go t s
+  where
+    go :: TTerm -> Set QName -> Set QName
+    go t qs = case t of
+      TDef q           -> Set.insert q qs
+      TApp f args      -> foldr go qs (f:args)
+      TLam b           -> go b qs
+      TCon q           -> Set.insert q qs
+      TLet a b         -> foldr go qs [a, b]
+      TCase _ _ p alts -> foldr qnamesInAlt (go p qs) alts
+      _                -> qs
+      where
+        qnamesInAlt a qs = case a of
+          TACon q _ t -> Set.insert q (go t qs)
+          TAGuard t b -> foldr go qs [t, b]
+          TALit _ b   -> go b qs
+
+
+-- | The argument allNames is optional. If you provide an empty list concrete
+-- names will not be appended to the NameId
+getCompilerEnv :: [[QName]] -> [(QName, TTerm)] -> TCM Mlf.Env
+getCompilerEnv allcons bs
+  | any ((>rangeSize Mlf.mlfTagRange) . length) allcons = error "too many constructors"
+  | otherwise = do
+      wa <- withArity allcons
+      return (Mlf.mkCompilerEnv2 allNames wa)
+  where
+    allNames = Set.toList $ foldr step mempty bs
+    step (qn, tt) acc = qnamesInTerm (Set.insert qn acc) tt
+
+-- | Creates a mapping for all the constructors in the array. The constructors
+-- should reference the same data-type.
+withArity :: [[QName]] -> TCM [[(QName, Int)]]
+withArity = mapM (mapM (\q -> (q,) <$> arityQName q))
+
+-- | If the qnames references a constructor the arity of that constructor is returned.
+arityQName :: QName -> TCM Int
+arityQName q = f . theDef <$> getConstInfo q
+  where
+    f def = case def of
+      Constructor{} -> conArity def
+      _             -> error "Not a constructor :("
+
+getBindings :: Definition -> TCM (Maybe (QName, TTerm))
+getBindings Defn{defName = q} = fmap (\t -> (q, t)) <$> toTreeless q
+
+mlfPostCompile :: MlfOptions -> IsMain -> Map ModuleName [Definition] -> TCM ()
+mlfPostCompile opts _ modToDefs = do
+  mapM_ (definitionSummary opts) allDefs
+  void $ mlfPostModule opts allDefs
+  where
+    allDefs :: [Definition]
+    allDefs = concat (Map.elems modToDefs)
+
+-- TODO: `Definition`'s should be sorted *and* grouped by `defMutual` (a field
+-- in Definition). each group should compile to:
+--
+--    (rec
+--       x0 = def0
+--       ...
+--    )
+mlfPostModule :: MlfOptions -> [Definition] -> TCM Mod
+mlfPostModule opts defs = do
+  modl@(MMod binds _) <- mlfMod defs
+  let modlTxt = prettyShow $ fromMaybe modl
+       ((withPrintInts modl . pure)  <$>  (_resultVar opts >>=  fromSimpleIdent binds))
+  when (_debug opts) $ liftIO . putStrLn $ modlTxt
+  whenJust (_resultVar opts) (printVars opts modl . pure)
```

Comment:
You maybe can use the Agda debug tracing infrastructure for some parts of this debug info. E.g. "reportSLn" and companions.